### ISSUE TITLE
Timeout (failed connection) when trying to access SQS from inside a Lambda

### ISSUE THREAD

**Issue by j6k4m8** [localstack/localstack#2875]
| | |
|---|---|
| Repository for minimal bug reproduction |https://github.com/j6k4m8/localstack-repro-2875
| Temporary workaround |When using `docker-compose`, hard-code `172.17.0.1` instead of using `$LOCALSTACK_HOSTNAME`. |

---

# Type of request: This is a ...

- [x] bug report
- [ ] feature request

# Detailed description

When trying to access SQS via (built-in, preinstalled) boto inside a lambda,

```python
import boto3

def main(event, lambda_context):

    sqs_client = boto3.client(
        "sqs",
        endpoint_url="http://" + os.getenv("LOCALSTACK_HOSTNAME") + ":4566",
        aws_access_key_id="foo",
        aws_secret_access_key="bar",
        region_name='us-east-1'
    )

    queue_url = sqs_client.get_queue_url(QueueName="MyFavoriteQueue")["QueueUrl"]
    ...
```

...the python execution hangs indefinitely on `get_queue_url` (or any other use of the client), until the lambda times out. The localstack logs say the following:

```
[ERROR] ConnectTimeoutError: Connect timeout on endpoint URL: "http://172.18.0.2:4566/"
```

This suggests that the `$LOCALSTACK_HOSTNAME` is being set correctly (or at least, is being set at all), but the service isn't reachable from inside the lambda.

## Expected behavior

Should be able to connect to SQS running on `$LOCALSTACK_HOSTNAME` from inside lambda

## Actual behavior

Cannot connect to SQS service; system times out

# Steps to reproduce

## Command used to start LocalStack

Clone the latest version of this repo, and then:

```shell
AWS_DEFAULT_REGION=us-east-1 \
AWS_PROFILE='localstack' \
AWS_ACCESS_KEY_ID='foobar' \
AWS_SECRET_ACCESS_KEY='foobar' \
TMPDIR=/private$TMPDIR \
DEBUG=1 \
SERVICES=serverless,cloudformation,sqs,events \
docker-compose up
```

This queue definitely exists (shows up in `awslocal sqs list-queues`), created like this: 

```python
self.sqs_client.create_queue(
    QueueName=queue_name, Attributes={"FifoQueue": "false"}
)
```

The lambda is created by mounting a zip file with the file above,

```python
lambda_function = self.lambda_client.create_function(
    FunctionName=function_name,
    Runtime="python3.8",
    Role=role_arn,
    Handler="my_file.main",
    Code={"ZipFile": my_exciting_bytes},
)
```

and then attaching an SQS event listener:

```python
lambda_client.create_event_source_mapping(
    EventSourceArn=queue_arn,
    FunctionName=lambda_arn,
    Enabled=True,
    BatchSize=1,
)
```

---

This issue seems very related to #2554, but the #2633 fix doesn't appear to work (I pulled the latest docker images just to be sure, but they're currently only a few hours old).

The lambda executes (and can be invoked directly), but hangs specifically on reaching out to the SQS service, so I'm not convinced it's an issue with the event source mapping.

Very grateful for your thoughts!

**Comment by whummer** [comment]
Thanks for reporting @j6k4m8 . This sounds like a networking issue in your Docker environment (potentially due to a local firewall or corporate VPN connection on your machine?). Which operating system are you on? Are you using the default docker-compose.yml from the repo, without any modifications (like Docker network settings, etc). Thanks

**Comment by j6k4m8** [comment]
Yikes, can't believe I completely forgot to include any of that.

Don't think it's a networking issue, but I will play around with that. I'm on macOS 10.15.6, so it's possible there's something funky going on with my docker process. This is indeed with the default, unmodified docker-compose.yml, with a pretty fresh install of docker as well. Encountering the same issue with `image: localstack/localstack:latest` (rather than just `localstack/localstack`).

**Comment by j6k4m8** [comment]
@duongpv7 and @whummer, thank you for your thoughts so far. I have created the following repository with a minimum working example to demonstrate this bug:

https://github.com/j6k4m8/localstack-repro-2875

**Comment by yonahforst** [comment]
I'm experiencing a similar issue with SNS.

Here's my `docker-compose.yml`
```yml
version: '2.1'

services:
  localstack:
    image: localstack/localstack
    restart: always
    ports:
      - "${EDGE_PORT-4566}:4566"
    environment:
      - DEFAULT_REGION=eu-west-1
      - USE_LIGHT_IMAGE=1
      - NODE_TLS_REJECT_UNAUTHORIZED=0
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"
      - "./startup_scripts:/docker-entrypoint-initaws.d"
      - "/codebase:/codebase"
```

and the error:
```
localstack_1  | START RequestId: 071f3436-647a-1675-4ba2-a466628e38a1 Version: $LATEST
localstack_1  | 2020-09-01T14:36:55.937Z	071f3436-647a-1675-4ba2-a466628e38a1	INFO	PAYLOAD: {"text":"yo","service_user_id":"https://125be8b5d114.ngrok.io"}
localstack_1  | 2020-09-01T14:36:55.939Z	071f3436-647a-1675-4ba2-a466628e38a1	INFO	ERROR: Error: connect ETIMEDOUT 172.29.0.2:4575
localstack_1  |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16) {
localstack_1  |   errno: 'ETIMEDOUT',
localstack_1  |   code: 'NetworkingError',
localstack_1  |   syscall: 'connect',
localstack_1  |   address: '172.29.0.2',
localstack_1  |   port: 4575,
localstack_1  |   region: 'eu-west-1',
localstack_1  |   hostname: '172.29.0.2',
localstack_1  |   retryable: true,
localstack_1  |   time: 2020-09-01T14:36:55.928Z
localstack_1  | }
localstack_1  | 2020-09-01T14:36:55.940Z	071f3436-647a-1675-4ba2-a466628e38a1	ERROR	
localstack_1  |   Error: connect ETIMEDOUT 172.29.0.2:4575
localstack_1  |       at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16)
localstack_1  | 
localstack_1  | END RequestId: 071f3436-647a-1675-4ba2-a466628e38a1
localstack_1  | REPORT RequestId: 071f3436-647a-1675-4ba2-a466628e38a1	Init Duration: 621.24 ms	Duration: 30000.00 ms	Billed Duration: 30000 ms	Memory Size: 1536 MBMax Memory Used: 68 MB	
```

**Comment by yonahforst** [comment]
I spent the last few days trying to debug this and here's where I landed:

**serverless.yml**
```serverless.yml
service: temp1

custom:
  defaultStage: local
  stage: ${opt:stage, self:custom.defaultStage}
  localstack:
    host: http://xx.xx.xx  ## ip of my ec2 instance
    stages:
      - local

provider:
  name: aws
  runtime: nodejs12.x
  stage: ${self:custom.stage}

functions:
  hello:
    handler: handler.hello

plugins:
  - serverless-localstack
```

**handler.js**
```js
const AWS = require('aws-sdk')

const sns = new AWS.SNS({
  region: 'us-east-1',
  endpoint: `http://${process.env.LOCALSTACK_HOSTNAME}:4566`,
})

module.exports.hello = event => {
  return sns.createTopic({ Name: 'foobar' }).promise()
};
```

**Using `localstack start`**
If I start localstack using the cli (`localstack start`), no problem. On the first invoke, there's an error in the the localstack logs (below) but the invocation works, and the topic is created

`awslocal lambda invoke --function-name=temp1-local-hello response.json`

```
2020-09-04T14:35:35:WARNING:localstack.services.cloudformation.cloudformation_starter: Unable to resolve "Ref" attribute for: {'Ref': 'HelloLambdaVersionEvclCFqSNll6luakPrqJaomIdi5TR7ZP6H6RWFm4'} - HelloLambdaVersionEvclCFqSNll6luakPrqJaomIdi5TR7ZP6H6RWFm4 - <class 'moto.awslambda.models.LambdaVersion'>
ERROR: 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal': exit code 1; output: b'\nError: No such object: ip-172-31-13-90.eu-west-1.compute.internal\n'
ERROR: 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal': exit code 1; output: b'\nError: No such object: ip-172-31-13-90.eu-west-1.compute.internal\n'
2020-09-04T14:35:53:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
2020-09-04T14:35:53:INFO:localstack.services.awslambda.lambda_api: Error executing Lambda function arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello: Command 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal' returned non-zero exit status 1. Traceback (most recent call last):
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 102, in get_main_endpoint_from_container
    DOCKER_MAIN_CONTAINER_IP = bootstrap.get_main_container_ip()
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 187, in get_main_container_ip
    container_name = get_main_container_name()
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 195, in get_main_container_name
    return run(cmd).strip().lstrip('/')
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 595, in run
    raise e
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 555, in run
    output = subprocess.check_output(cmd, shell=True, stderr=stderr, env=env_dict, cwd=cwd)
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/code/localstack/localstack/services/awslambda/lambda_api.py", line 530, in run_lambda
    result = LAMBDA_EXECUTOR.execute(func_arn, func_details, event, context=context,
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 167, in execute
    return do_execute()
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 159, in do_execute
    return _run(func_arn=func_arn)
  File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 146, in wrapped
    raise e
  File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 142, in wrapped
    result = func(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 152, in _run
    raise e
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 140, in _run
    result = self._execute(func_arn, func_details, event, context, version)
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 252, in _execute
    main_endpoint = get_main_endpoint_from_container()
  File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 105, in get_main_endpoint_from_container
    container_name = bootstrap.get_main_container_name()
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 195, in get_main_container_name
    return run(cmd).strip().lstrip('/')
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 595, in run
    raise e
  File "/opt/code/localstack/localstack/utils/bootstrap.py", line 555, in run
    output = subprocess.check_output(cmd, shell=True, stderr=stderr, env=env_dict, cwd=cwd)
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal' returned non-zero exit status 1.

2020-09-04T14:36:12:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
2020-09-04T14:36:12:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.2891f7e4/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
```

**Using docker-compose**

If I start localstack using `docker-compose` (using the docker-compose.yml from [master](https://github.com/localstack/localstack/blob/master/docker-compose.yml))

the same invocation `awslocal lambda invoke --function-name=temp1-local-hello response.json` eventually times-out:
```
localstack_main | 2020-09-04T14:37:49:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
localstack_main | 2020-09-04T14:37:49:INFO:localstack.services.awslambda.lambda_executors: Determined main container target IP: 172.19.0.2
localstack_main | 2020-09-04T14:37:49:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.b576b953/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
localstack_main | 2020-09-04T14:38:50:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
localstack_main | 2020-09-04T14:38:50:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.b576b953/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
localstack_main | 2020-09-04T14:39:51:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
localstack_main | 2020-09-04T14:39:51:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.b576b953/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
localstack_main | 2020-09-04T14:40:54:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
localstack_main | 2020-09-04T14:40:54:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.b576b953/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
localstack_main | 2020-09-04T14:42:02:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello"
localstack_main | 2020-09-04T14:42:02:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: CONTAINER_ID="$(docker create -i   -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"  --rm "lambci/lambda:nodejs12.x" "handler.hello")";docker cp "/tmp/localstack/zipfile.b576b953/." "$CONTAINER_ID:/var/task"; docker start -ai "$CONTAINER_ID";
localstack_main | 2020-09-04T14:46:36:INFO:localstack.services.awslambda.lambda_api: Error executing Lambda function arn:aws:lambda:us-east-1:000000000000:function:temp1-local-hello: Lambda process returned error status code: 1. Result: {"errorType":"NetworkingError","errorMessage":"connect ETIMEDOUT 172.19.0.2:4566"}. Output:
localstack_main | START RequestId: 816aa762-a471-17c9-0fb7-7ba8c6c33b32 Version: $LATEST
localstack_main | 2020-09-04T14:46:35.689Z	816aa762-a471-17c9-0fb7-7ba8c6c33b32	ERROR	Invoke Error 	{"errorType":"NetworkingError","errorMessage":"connect ETIMEDOUT 172.19.0.2:4566","code":"NetworkingError","message":"connect ETIMEDOUT 172.19.0.2:4566","errno":"ETIMEDOUT","syscall":"connect","address":"172.19.0.2","port":4566,"region":"us-east-1","hostname":"172.19.0.2","retryable":true,"time":"2020-09-04T14:46:35.688Z","stack":["Error: connect ETIMEDOUT 172.19.0.2:4566","    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16)"]}
localstack_main | END RequestId: 816aa762-a471-17c9-0fb7-7ba8c6c33b32
localstack_main | REPORT RequestId: 816aa762-a471-17c9-0fb7-7ba8c6c33b32	Init Duration: 333.88 ms	Duration: 6000.00 ms	Billed Duration: 6000 ms	Memory Size: 1536 MB	Max Memory Used: 26 MB	 Traceback (most recent call last):
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_api.py", line 530, in run_lambda
localstack_main |     result = LAMBDA_EXECUTOR.execute(func_arn, func_details, event, context=context,
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 167, in execute
localstack_main |     return do_execute()
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 159, in do_execute
localstack_main |     return _run(func_arn=func_arn)
localstack_main |   File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 146, in wrapped
localstack_main |     raise e
localstack_main |   File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 142, in wrapped
localstack_main |     result = func(*args, **kwargs)
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 152, in _run
localstack_main |     raise e
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 140, in _run
localstack_main |     result = self._execute(func_arn, func_details, event, context, version)
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 297, in _execute
localstack_main |     result = self.run_lambda_executor(cmd, stdin, env_vars=environment, func_details=func_details)
localstack_main |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 204, in run_lambda_executor
localstack_main |     raise Exception('Lambda process returned error status code: %s. Result: %s. Output:\n%s' %
localstack_main | Exception: Lambda process returned error status code: 1. Result: {"errorType":"NetworkingError","errorMessage":"connect ETIMEDOUT 172.19.0.2:4566"}. Output:
localstack_main | START RequestId: 816aa762-a471-17c9-0fb7-7ba8c6c33b32 Version: $LATEST
localstack_main | 2020-09-04T14:46:35.689Z	816aa762-a471-17c9-0fb7-7ba8c6c33b32	ERROR	Invoke Error 	{"errorType":"NetworkingError","errorMessage":"connect ETIMEDOUT 172.19.0.2:4566","code":"NetworkingError","message":"connect ETIMEDOUT 172.19.0.2:4566","errno":"ETIMEDOUT","syscall":"connect","address":"172.19.0.2","port":4566,"region":"us-east-1","hostname":"172.19.0.2","retryable":true,"time":"2020-09-04T14:46:35.688Z","stack":["Error: connect ETIMEDOUT 172.19.0.2:4566","    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16)"]}
localstack_main | END RequestId: 816aa762-a471-17c9-0fb7-7ba8c6c33b32
```

**Analysis:**
When localstack is started via the CLI, we see an error `Command 'docker inspect -f '{{ .Name }}' ip-172-31-13-90.eu-west-1.compute.internal' returned non-zero exit status 1` this gives us a hint, because the last argument should be a container ID (not whatever that is) so it's is probably related to this commit: https://github.com/localstack/localstack/commit/5b9fe560d302ad97259a850a4fff229f96ca0aee#

I think this is what's happening: In the commit mentioned above, we expect the env variable $HOSTNAME to be the container ID. But on ec2 it's set to the instances private dns and docker doesn't overwrite it (i guess), so the trick we're using in that commit to get the container name using the $HOSTNAME variable doesn't work.

The reason we don't get this error using docker-compose is because we explicitly specify which env variables should be available to the container, and HOSTNAME isn't one of them, so docker sets it to the container ID and the trick works.

Ok, but now why does this bug actually fix the timeout issue for me?

I think what's happening is that the error above is caught [here](https://github.com/localstack/localstack/blob/master/localstack/services/awslambda/lambda_executors.py#L109) and the endpoint defaults to DOCKER_HOST_FROM_CONTAINER which is set from DOCKER_BRIDGE_IP which defaults to '172.17.0.1'

Indeed, if I change my code and use a hardcoded endpoint, like so:
```js
const sns = new AWS.SNS({
  region: 'us-east-1',
  endpoint: `http://172.17.0.1:4566`,
})

module.exports.hello = event => {
  return sns.createTopic({ Name: 'foobar' }).promise()
};
```
it works!

I'm not sure how or why, but at least I work around for the moment. I don't really understand the network setup and why one works but not the other, so I'll leave that to you guys <3 . In the meantime I'll see if I can find a better way to get the main container name which doesnt break on ec2.

**Comment by j6k4m8** [comment]
I can confirm that @yonahforst's fix also works on my repro codebase ([comparison here](https://github.com/j6k4m8/localstack-repro-2875/pull/2)). 
cc @duongpv7

**Comment by whummer** [comment]
Thanks for the detailed analysis @yonahforst . Indeed, somehow could be related to related to this commit: 5b9fe56 

One thing I don't fully understand: what is the intention of this line in serverless.yml ? 
```
    host: http://xx.xx.xx  ## ip of my ec2 instance
```

I think the name `host` may be misleading - this configuration value basically maps to the `LOCALSTACK_HOSTNAME` that will be used to make requests to (by default simply on `localhost`). Could you simply use `http://localhost` here, or even leave out this configuration and use the default value?

**Comment by whummer** [comment]
Also, please note that your client code apparently tries to connect to port `4575`, whereas only the edge port `4566` is exposed in Docker (which should be the correct setup).

**Comment by j6k4m8** [comment]
Thanks for the pointers @whummer! Can confirm that this error persists even with [the default compose yaml](https://github.com/j6k4m8/localstack-repro-2875#usage) and with [:4566 specified in the container](https://github.com/j6k4m8/localstack-repro-2875/blob/master/create-resources.py#L43).

**Comment by yonahforst** [comment]
Thanks @whummer! The I'm using the host configuration because I'm running localstack on an EC2 server and deploying from my local machine. the `http://xx.xx.xx` is replaced with the actual ip address of my server.

In the first example, yes, the client tries to connect on 4575 but in the second example, it uses 4566. In fact, the client code is being executed within a lambda container within the localstack docker, so it should still have access to (the deprecated) port 4575 (and it did in previous builds).

I guess the main point is that: within the localstack docker, we can access AWS services at 172.17.0.1 but not 172.19.0.2

**Comment by yonahforst** [comment]
I've made a bit of progress here: I think what's happening is that the lambda containers and the localstack container are running on different networks. Here's a diagram I generated using [this tool](https://github.com/LeoVerto/docker-network-graph)
![Untitled 2](https://user-images.githubusercontent.com/1440796/93440997-ee690480-f8cf-11ea-8e2e-277178922600.png)
Note: the IP addresses are different here. It seems that docker (or docker-compose, dunno) automatically assigns them, so 172.17.0.1 won't work if docker assigned your network a different IP.

So why does 172.17.0.1 work? I'm not sure. I guess it's because it's the gateway ip? and somehow that allows other containers running on the BRIDGE network to access it?

OH, this just clicked for me. the lambda containers running on the bridge network have access to the EC2 instance via the gateway IP. The localstack container maps to port 4566 on the EC2 instance. So when you call the gateway IP it goes to the EC2 instance which then forwards it to the localstack container. Durrrrr!!

So, I'm not sure what the correct behavior should be, but I managed to work around it by setting the network mode to bridge in my docker-compose.yml, so that all containers are on the same network:
```yml
version: '2.1'

services:
  localstack:
    image: localstack/localstack
    network_mode: bridge
    ports:
      - "${EDGE_PORT-4566}:4566"
    environment:
      - DEFAULT_REGION=eu-west-1
      - NODE_TLS_REJECT_UNAUTHORIZED=0
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"
``` 
now my network diagram looks like this:
![Untitled 2](https://user-images.githubusercontent.com/1440796/93452432-b3b79a00-f8d8-11ea-8414-1c222c7749f9.png)
and things work! (note: I'm running several instances of localstack on the same machine, that's why there are so many localstack containers)
